### PR TITLE
Add OpenGL renderer

### LINKED ISSUE #877: Use OpenGL 3.3 for the low-end rendering backend
_edit: see https://github.com/godotengine/godot-proposals/issues/877#issuecomment-633826826 below for updates. Make sure to read both this post and the updates before responding._

**Describe the project you are working on:**
The Godot engine renderer

## The Problem
**Describe the problem or limitation you are having in your project:**

Implementing new 3D features is becoming increasingly difficult as we try to support modern workflows (e.g. HDR, PBR) while maintaining support for OpenGL 2.1/ES 2.0 devices. 

We frequently have to sacrifice quality in our high-end rendering backend in order to support consistent workflows between the high-end (Vulkan) and low-end (OpenGL) backends.

Some common issues that arise:
1. Scenes look very different in GLES3 vs GLES2 because of the use of sRGB vs linear lighting
2. Many optimizations are not possible in GLES2
3. We have to rewrite many parts of core rendering code to support GLES2 devices of varying capabilities
5. Many common post processing and material effects are just not possible in GLES2
6. Common 3D engine features can't work in GLES2, so we don't support them (e.g. terrain, complex particle systems)
7. Our rendering defaults have to be compatible with GLES2 and GLES3, leading to poor quality default scenes overall

### Compatibility with older Devices
Right now the GLES2 backend serves 2 purposes:
1. Compatibility with older devices that only support GLES2
2. Providing a lower-end renderer for lower-end devices and for users who value speed over modern features

When we decided to add the GLES2 backend many devices only supported GLES2. At the time, 40% of Android devices only supported GLES2. Thankfully that is no longer true.

OpenGL 2.1/ ES 2.0 only devices are becoming increasingly rare. On Android over [85% of devices now support GL ES 3.0  or higher](https://developer.android.com/about/dashboards#OpenGL). On iOS, every device supports ES 3.0. On Desktop the overwhelming majority of devices support GL 3.3+ (more than 99% support at least GL 3.3 according to the [Steam hardware survery](https://store.steampowered.com/hwsurvey/Steam-Hardware-Software-Survey-Welcome-to-Steam))

As a result of the rapid adoption of OpenGL/ES 3.x+ the main purpose served by the GLES2 renderer is as a low-end renderer to maximize speed on high and low-end devices. Many users that are choosing to use GLES2 currently are not doing so because their device doesn't support OpenGL 3.3, they are doing so because their device doesn't like the high-end features that are in the Godot GLES3 renderer.

## The Solution
**Describe the feature / enhancement and how it helps to overcome the problem or limitation:**
I propose dropping GLES2 support for 3D. For 2D, we can still use GLES2, but when 3D features are used, we will require GLES3.

Given that the main purpose behind the OpenGL backend will be supporting low-end devices, we will write the GLES3 renderer to target low-end devices. This means it won't have the feature set that that Godot 3.2 GLES3 renderer has, but it should run even faster than the current GLES2 renderer and look much better.

This change will both increase the perceived speed and quality of the low-end renderer, but it will also make development much smoother.

What this allows us to:
1. Standardize workflow between high-end (Vulkan) and low end (OpenGL) backends. This means you can expect a more uniform look when switching between backends.
2. Implement optimizations not present in GLES2 (using instancing, texture arrays, and transform feedback)
3. Have similar lightmap workflow between backends
4. Greatly simplify our engine code which makes maintaining the engine easier
5. Extend more post-processing/materials to the OpenGL backend
6. Support more high-end features that have become standard in 3D engines (e.g. terrain, physical light units, HDR)

The one issue with this proposal is WebGL. WebGL 2.0 is not supported well yet. We will either have to work out a sort of compatibility mode for WebGL, or wait for more widespread adoption of the WebGL 2.0 API.

If we do drop 3D support for GLES2 in 4.0, it will mean that we will have to work harder to support 3.2.x for a long time. Users will still be able to make and export games using Godot 3.2.x and critical bugfixes will still be merged for 3.2.x. This seems to be a fair trade off. If the GLES2 renderer is ported to Godot 4.0, it won't have access to any of the additional features that come with 4.0.

One of the factors we considered when moving to Vulkan was the poor quality of OpenGL ES 3.0 drivers on mobile. At this point, OpenGL ES 3.0 drivers support seems to have improved substantially. Further, basic features are supported very well in the OpenGL ES 3.0 drivers. By limiting ourselves to a limited feature set (closer to GLES2) then we won't bump up against any of the features that are poorly supported by the drivers and we should be able to avoid any of the major problems we had before. 

**Describe how your proposal will work, with code, pseudocode, mockups, and/or diagrams:**
_This section is edited based on comments below_

User will be able to choose between 4 backends:
1. High-end (Vulkan)
2. Low-end/mobile (Vulkan)
3. GLES3 (more limited than current GLES3 backend, similar feature set to current GLES2 backend. Aims for maximum compatibility)
4. GLES2 (2D only to start, may support very basic 3D eventually)

**If this enhancement will not be used often, can it be worked around with a few lines of script?:**
No. 

**Is there a reason why this should be core and not an add-on in the asset library?:**
Yes, this has to do with core engine functionality

### PR DESCRIPTION
**Pull request by Calinou**
Based on @lawnjelly's efforts on the GLES2 renderer here: https://github.com/godotengine/godot/pull/44399 :slightly_smiling_face:

Despite the [news article stating that OpenGL support will return in Godot 4.1](https://godotengine.org/article/about-godot4-vulkan-gles3-and-gles2), this pull request is targeting a merge for Godot 4.0. However, **we don't expect this OpenGL renderer to be fully complete (especially on the 3D side) when Godot 4.0 is released**. Nonetheless, it may still be sufficient for many 2D games and non-game applications that want to maximize compatibility.

Commits are unsquashed for now to ease troubleshooting and tracking regressions. Most of them should be squashed before merging this PR.

This closes https://github.com/godotengine/godot-proposals/issues/877.

### OpenGL renderer's advantages over Vulkan

- Much better compatibility with old devices (desktop, mobile).
- Can run on HTML5 using WebGL 2.0 once the platform layer is reimplemented.
  - In contrast, [WebGPU isn't available in all stable browsers yet](https://caniuse.com/webgpu). Moreover, a RenderingDevice abstraction needs to be written for it.
- Faster startup times thanks to simpler shaders (even with the lack of shader caching).

*When targeting desktop platforms, using Vulkan is still recommended for several reasons:*

### Vulkan renderer's advantages over OpenGL

- Better visuals and performance in complex scenes.
  - If you need a lower base cost for integrated graphics or low-end dedicated GPUs, this can be achieved by using the Vulkan Mobile backend on desktop platforms.
- No issues with AMD GPU performance on Windows (due to subpar OpenGL implementation).

Nonetheless, for mobile platforms, I expect OpenGL to be more prevalent for a while.

### Differences from [lawnjelly's pull request](https://github.com/godotengine/godot/pull/44399)

- OpenGL 3.3 / OpenGL ES 3.0 / WebGL 2.0 is now used as a baseline. **The renderer will still follow a low-end-friendly approach** to maximize performance and compatibility on low-end devices.
- Removed the `gles_common` abstraction as there will only be one OpenGL renderer.
- "GLES2" renderer renamed to "OpenGL" for consistency with the Vulkan renderer, which does not specify a version in its name. Use `--rendering-driver opengl` to force a project to start with the OpenGL renderer, or use the dropdown in the top-right corner of the editor.

### Current limitations

- **2D only.** No 3D rendering support for now.
  - Getting 2D rendering support to be reliable for daily use is the current priority, as it'll unblock contributors who don't have a GPU that supports Vulkan.
- Only the Linux/X11 context is currently implemented. Due to this, the OpenGL renderer can't be used on Windows, macOS, Android, iOS and HTML5 yet.

### TODO

- 2D editor viewport rendering is currently broken, but the running project displays correctly.
- The OpenGL Core profile should be used instead of the Compatibility profile. The compatibility profile was temporarily enabled to allow Godot to render correctly after switching from a GLES2 context to a GLES3 context.
- Shaders should be ported from GLSL ES 1.0 to GLSL ES 3.0 using the Vulkan GLSL shaders as a base.
- Some texture functions need to be reimplemented.
- The buildsystem code should be checked to make sure OpenGL and Vulkan can be disabled individually at compile-time. This lets people build smaller binaries when needed.
- The shader compiler needs to be rewritten to use a modern approach. Ubershaders and asynchronous compilation will likely be used to avoid stutters during gameplay. This will be handled by @reduz and @RandomShaper.
- Decide what to do with the `GLES_ENABLED` and `OPENGL_BACKEND_ENABLED` defines. Should we merge them with `OPENGL_ENABLED` to be consistent with `VULKAN_ENABLED`?
- rasterizer_array.h should be replaced with LocalVector: https://github.com/Calinou/godot/blob/add-opengl-renderer/drivers/opengl/rasterizer_array.h#L37

## Preview

Godot editor running on OpenGL

### CODE REVIEW

**Comment by clayjohn** [comment]
A couple notes:
1. Have you tested on Windows? I had GLES3 working in Windows on Lawnjelly's branch earlier, I imagine that it should be working or close to working still
2. The rasterizer should not depend on the dummy rasterizer at all. For now, it should be enough to copy-paste the relevant sections of the dummy into the relevant files in drivers/GLES3

File: core/core_bind.h [hunk 1]
```diff
@@ -119,8 +119,8 @@ class OS : public Object {
 
 public:
 	enum VideoDriver {
-		VIDEO_DRIVER_GLES2,
 		VIDEO_DRIVER_VULKAN,
+		VIDEO_DRIVER_OPENGL,
```
Comment (aaronfranke) [inline core/core_bind.h:123]: `O` comes before `V` in the alphabet, there's no purpose to swapping the order here.

**Comment by qarmin** [comment]
I think that it should be temporary to CI to see progress by adding `--rendering-driver opengl` to this commands
https://github.com/godotengine/godot/blob/8abd46096ae52025d54bf58a782a6c057b3254f6/.github/workflows/linux_builds.yml#L115-L127

**Comment by qarmin** [comment]
On Intel HD 3000 and Mesa 21.3, I see empty screen in both project manager and editor
Zrzut ekranu z 2021-09-28 08-46-27
and also I see this errors
```
Mesa: User error: GL_INVALID_OPERATION in glDrawArrays
Mesa: User error: GL_INVALID_OPERATION in glVertexAttribPointer(no array object bound)
Mesa: User error: GL_INVALID_OPERATION in glDrawArrays
Mesa: User error: GL_INVALID_OPERATION in glVertexAttribPointer(no array object bound)
Mesa: User error: GL_INVALID_OPERATION in glVertexAttribPointer(no array object bound)
Mesa: User error: GL_INVALID_OPERATION in glVertexAttribPointer(no array object bound)
Mesa: User error: GL_INVALID_OPERATION in glDrawElements
Mesa: User error: GL_INVALID_OPERATION in glVertexAttribPointer(no array object bound)
Mesa: User error: GL_INVALID_OPERATION in glDrawArrays
Mesa: User error: GL_INVALID_VALUE in glTexImage2D(internalFormat=GL_LUMINANCE_ALPHA)
Mesa: User error: GL_INVALID_VALUE in glTexImage2D(internalFormat=GL_LUMINANCE_ALPHA)
```

File: core/core_bind.h (hunk 1 above)
Comment (Calinou) [reply core/core_bind.h:123]: I swapped the order to match the order found in the project settings' enum: https://github.com/Calinou/godot/blob/8fb7cafaeb31b9d27b199effbc8a1829a0fa4845/main/main.cpp#L1271

**Comment by jordo** [comment]
Is Godot 4.0 launching without html5/web support then?

**Comment by Calinou** [comment]
> Is Godot 4.0 launching without html5/web support then?

It depends on whether we can finish this OpenGL renderer (+ the WebGL platform port) in time. Either way, `3.x` will remain supported until 4.1 is released.

**Comment by jordo** [comment]
> It depends on whether we can finish this OpenGL renderer (+ the WebGL platform port) in time for 4.0's release. We can't promise anything yet. Either way, `3.x` will remain supported until 4.1 is released.
> 
> WebGPU isn't likely to be implemented until 4.1 at least, so OpenGL via WebGL is our only option for HTML5 in the short term.


Would love a concise answer.  I understand 3.x is supported, but my question was specific to 4.0.  I'm not looking for promises, it would just be nice to be clear about this so developers can plan accordingly.  From other contributer comments, it seems Vulkan is going to be the only supported initial backend for 4.0.  Connecting the dots would therefore mean that web export won't be supported in 4.0 correct?  WebGPU isn't a viable option right now, nor the relatively near future.

I think this is an important platform to confirm this one way or the other.

**Comment by jordo** [comment]
I guess the PR of the actual renderer impl isn't the best place to ask these questions, but it would be nice to know if the plan is to have this available in 4.0 or not.  It's fine either way, just nice to know so people can plan and adjust accordingly.

**Comment by Calinou** [comment]
> Connecting the dots would therefore mean that web export won't be supported in 4.0 correct?

Indeed, web export (and web editor) won't be available until some way to render graphics is implemented in the `master` branch.

**Comment by Calinou** [comment]
> 1. Have you tested on Windows? I had GLES3 working in Windows on Lawnjelly's branch earlier, I imagine that it should be working or close to working still

It seems to crash on startup, but Vulkan also crashes in the same branch anyway. I haven't investigated why yet.

**Comment by clayjohn** [comment]
This is a good start! I was able to get it running on Windows (PR here https://github.com/Calinou/godot/pull/6)

A couple of things we need to take a look at still:
1. It looks like viewports are not working
2. The renderer is still using deprecated functionality not available in OpenGL 3.3, we should transition to OpenGL 3.3 fully before merging (can can likely copy-paste a lot from the GLES3 drivers)
3. Multiple windows are not working (and single window mode is broken because of 1)

**Comment by clayjohn** [comment]
Okay, I've submitted a pull request that removes much of the deprecated GLES2 functions (enough to get core profile working anyway). I have also fixed the bug that made viewports not display. https://github.com/Calinou/godot/pull/7

So right now we have a minimal 2D renderer in OpenGL 3.3 on Windows. @Calinou If you can set the X11 platform to core profile and verify that it works there as well then I think we have a minimal working example here!

**Pull request closed without merging**